pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PublishRateLimiter.java

```diff
@@ -66,6 +69,83 @@
      * @param clusterName
      */
     void update(PublishRate maxPublishRate);
+
+    /**
+     * try to acquire permit
+     * @param numbers
+     * @param bytes
+     * */
+    boolean tryAcquire(int numbers, long bytes);
+}
+
+class PrecisPublishLimiter implements PublishRateLimiter {
+    protected volatile int publishMaxMessageRate = 0;
```

Comment:
I'd say in general to avoid volatile in all cases in which the exact semantic of read after write consistency is required. In this case we can probably just rely on dirty reads

Reply:
@merlimat  There will be barely concurrent update of these fields, so I think volatile here is ok.  Do you suggest to remove it ?